src/gin/nccl_ofi_gin_reqs.cpp

```diff
@@ -0,0 +1,159 @@
+		req, err_entry->err,
+		err_entry->prov_errno,
+		fi_cq_strerror(cq, err_entry->prov_errno, err_entry->err_data, NULL, 0),
+		(long)err_entry->len);
+
+	/*
+	 * Libfabric error codes directly map to ISO C errno values for standard
+	 * error codes up to FI_ERRNO_OFFSET, and libfabric-specific error codes
+	 * beyond. nccl_net_ofi_retval_translate() will figure out
+	 * how to deal with these, so it is safe to pass up the err as-is.
+	 * However, any special-handling for prov_errno should be handled here.
+	 */
+	ret = -(err_entry->err);
+	return ret;
+}
+
+int nccl_net_ofi_gin_writeack_req_t::handle_cq_entry(nccl_net_ofi_context_t *_ctx, fi_cq_entry *cq_entry_base, fi_addr_t src_addr, uint16_t rail_id)
+{
+	assert(gin_comm->outstanding_ack_counter > 0);
+	gin_comm->outstanding_ack_counter--;
+
+	delete this;
+	return 0;
+}
+
+nccl_net_ofi_gin_recv_req_t::nccl_net_ofi_gin_recv_req_t(nccl_ofi_gin_ep_t *gin_ep_arg,
+				    			 nccl_ofi_gin_ep_rail_t *rail_arg) :
+		nccl_net_ofi_gin_req_t(),
+		gin_ep(gin_ep_arg),
+		rail(rail_arg)
+{
+	rx_buff_elem = nccl_ofi_freelist_entry_alloc(gin_ep->rx_buff_fl.get());
+	if (!rx_buff_elem) {
+		NCCL_OFI_WARN("Failed to allocate rx buffer freelist entry");
+		throw std::runtime_error("Failed to allocate rx buffer freelist entry");
+	}
+}
+
+nccl_net_ofi_gin_recv_req_t::~nccl_net_ofi_gin_recv_req_t()
+{
+	nccl_ofi_freelist_entry_free(gin_ep->rx_buff_fl.get(), rx_buff_elem);
+}
+
+int nccl_net_ofi_gin_recv_req_t::handle_cq_entry(nccl_net_ofi_context_t *_ctx,
+						 fi_cq_entry *cq_entry_base, fi_addr_t src_addr,
+						 uint16_t rail_id_arg)
+{
+	assert(this->rail->rail_id == rail_id_arg);
+
+	auto *cq_entry = reinterpret_cast<struct fi_cq_data_entry *>(cq_entry_base);
+
+	auto *rdma_domain = gin_ep->domain;
+
+	int ret = 0;
+
+	if (cq_entry->flags & FI_REMOTE_WRITE) {
+		/* RDMA write-immediate completion */
+		uint32_t comm_id = GET_COMM_ID_FROM_IMM(cq_entry->data);
+
+		auto *gin_comm = static_cast<nccl_ofi_gin_comm_t *>(
+			rdma_domain->rdma_domain_get_device()->
+				rdma_device_get_comm(comm_id)
+		);
+
+		uint16_t msg_seq_num = GET_SEQ_NUM_FROM_IMM(cq_entry->data);
+		uint64_t total_segms = GET_NUM_SEG_FROM_IMM(cq_entry->data);
+		size_t len = cq_entry->len;
+
+		ret = gin_handle_signal_write_completion
+			(gin_comm, src_addr, rail_id_arg, msg_seq_num, total_segms, len);
+		if (ret != 0) {
+			return ret;
+		}
+	} else {
+		auto *msg = static_cast<nccl_net_ofi_rdma_signal_metadata_msg_t *>(rx_buff_elem->ptr);
+
+		/* Get the gin comm */
+		auto *gin_comm = static_cast<nccl_ofi_gin_comm_t *>(
+			rdma_domain->rdma_domain_get_device()->
+				rdma_device_get_comm(msg->remote_comm_id)
+		);
+
+		if (gin_comm == nullptr) {
+			NCCL_OFI_WARN("Failed to get gin comm");
+			return -EINVAL;
+		}
+
+		ret = gin_handle_signal_metadata_completion
+			(gin_comm, src_addr, rail_id_arg, msg);
+		if (ret != 0) {
+			NCCL_OFI_WARN("gin_handle_signal_metadata_completion failure");
+			return ret;
+		}
+	}
```

Comment:
The input to this is `rx_buff_elem` so does not depend on the arguments to this function. Maybe place this in a different function and have the caller decide to call this or that.

Reply:
Good to resolve.